File: src/ios/FantasyGMAssistant/NativeModules/Authentication/FirebaseAuthManager.m
//
//  FirebaseAuthManager.m
//  FantasyGMAssistant
//
//  Enhanced Firebase Authentication bridge with security features
//  Version: 1.0.0
//

#import "FirebaseAuthManagerBridge.h"
#import <React/RCTBridgeModule.h>
#import <FirebaseAuth/FirebaseAuth.h>
#import <LocalAuthentication/LocalAuthentication.h>
#import <CommonCrypto/CommonDigest.h>

// Constants for rate limiting and security
static const NSInteger kMaxAuthAttempts = 5;
static const NSTimeInterval kRateLimitResetInterval = 300.0; // 5 minutes
static const NSTimeInterval kTokenRefreshThreshold = 300.0; // 5 minutes before expiry

@implementation FirebaseAuthManager {
    NSInteger rateLimitCounter;
    NSDate *lastAuthAttempt;
    NSString *deviceFingerprint;
    dispatch_queue_t securityQueue;
}

RCT_EXPORT_MODULE(FirebaseAuthManager)

- (instancetype)init {
    if (self = [super init]) {
        rateLimitCounter = 0;
        lastAuthAttempt = [NSDate date];
        securityQueue = dispatch_queue_create("com.fantasygm.auth.security", DISPATCH_QUEUE_SERIAL);
        [self generateDeviceFingerprint];
    }
    return self;
}

#pragma mark - Security Methods

- (void)generateDeviceFingerprint {
    NSString *deviceName = [[UIDevice currentDevice] name];
    NSString *systemVersion = [[UIDevice currentDevice] systemVersion];
    NSString *model = [[UIDevice currentDevice] model];
    NSString *identifierForVendor = [[[UIDevice currentDevice] identifierForVendor] UUIDString];
    
    NSString *rawFingerprint = [NSString stringWithFormat:@"%@_%@_%@_%@",
                               deviceName, systemVersion, model, identifierForVendor];
    
    // Generate SHA-256 hash of device info
    NSData *data = [rawFingerprint dataUsingEncoding:NSUTF8StringEncoding];
    uint8_t digest[CC_SHA256_DIGEST_LENGTH];
    CC_SHA256(data.bytes, (CC_LONG)data.length, digest);
    
    NSMutableString *fingerprint = [NSMutableString stringWithCapacity:CC_SHA256_DIGEST_LENGTH * 2];
    for (int i = 0; i < CC_SHA256_DIGEST_LENGTH; i++) {
        [fingerprint appendFormat:@"%02x", digest[i]];
    }
    
    deviceFingerprint = fingerprint;
}

- (BOOL)checkRateLimit {
    __block BOOL allowed = NO;
    dispatch_sync(securityQueue, ^{
        NSTimeInterval timeSinceLastAttempt = [[NSDate date] timeIntervalSinceDate:self->lastAuthAttempt];
        
        if (timeSinceLastAttempt >= kRateLimitResetInterval) {
            self->rateLimitCounter = 0;
        }
        
        if (self->rateLimitCounter < kMaxAuthAttempts) {
            self->rateLimitCounter++;
            self->lastAuthAttempt = [NSDate date];
            allowed = YES;
        }
    });
    return allowed;
}

#pragma mark - Authentication Methods

RCT_EXPORT_METHOD(signInWithEmail:(NSString *)email
                  password:(NSString *)password
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
    
    if (![self checkRateLimit]) {
        NSString *errorMsg = @"Too many authentication attempts. Please try again later.";
        reject(@"AUTH_RATE_LIMIT", errorMsg, nil);
        return;
    }
    
    [[FIRAuth auth] signInWithEmail:email
                          password:password
                        completion:^(FIRAuthDataResult * _Nullable authResult,
                                   NSError * _Nullable error) {
        if (error) {
            [self logAuthError:error withEmail:email];
            reject(@"AUTH_ERROR", error.localizedDescription, error);
            return;
        }
        
        [self handleSuccessfulAuth:authResult withResolve:resolve reject:reject];
    }];
}

RCT_EXPORT_METHOD(signInWithProvider:(NSString *)provider
                  options:(NSDictionary *)options
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
    
    if (![self checkRateLimit]) {
        reject(@"AUTH_RATE_LIMIT", @"Too many authentication attempts", nil);
        return;
    }
    
    // Configure provider
    id<FIRAuthProvider> authProvider;
    if ([provider isEqualToString:@"google"]) {
        authProvider = [FIRGoogleAuthProvider provider];
    } else if ([provider isEqualToString:@"espn"]) {
        // Custom ESPN OAuth provider configuration
        authProvider = [self configureESPNProvider:options];
    } else {
        reject(@"AUTH_PROVIDER_ERROR", @"Unsupported provider", nil);
        return;
    }
    
    [[FIRAuth auth] signInWithProvider:authProvider
                           UIDelegate:nil
                           completion:^(FIRAuthDataResult * _Nullable authResult,
                                      NSError * _Nullable error) {
        if (error) {
            [self logAuthError:error withProvider:provider];
            reject(@"AUTH_ERROR", error.localizedDescription, error);
            return;
        }
        
        [self handleSuccessfulAuth:authResult withResolve:resolve reject:reject];
    }];
}

RCT_EXPORT_METHOD(signOut:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
    
    NSError *error;
    [[FIRAuth auth] signOut:&error];
    
    if (error) {
        [self logAuthError:error withContext:@"signOut"];
        reject(@"SIGNOUT_ERROR", error.localizedDescription, error);
        return;
    }
    
    // Clear security state
    dispatch_sync(securityQueue, ^{
        self->rateLimitCounter = 0;
        self->lastAuthAttempt = [NSDate date];
    });
    
    resolve(@{@"status": @"success"});
}

RCT_EXPORT_METHOD(getCurrentUser:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
    
    FIRUser *currentUser = [FIRAuth auth].currentUser;
    if (!currentUser) {
        resolve(nil);
        return;
    }
    
    [currentUser getIDTokenForcingRefresh:NO
                              completion:^(NSString * _Nullable token,
                                         NSError * _Nullable error) {
        if (error) {
            [self logAuthError:error withContext:@"tokenRefresh"];
            reject(@"TOKEN_ERROR", error.localizedDescription, error);
            return;
        }
        
        [self validateAndEnhanceUserData:currentUser
                              withToken:token
                               resolve:resolve
                                reject:reject];
    }];
}

#pragma mark - Helper Methods

- (void)handleSuccessfulAuth:(FIRAuthDataResult *)authResult
                withResolve:(RCTPromiseResolveBlock)resolve
                   reject:(RCTPromiseRejectBlock)reject {
    
    [authResult.user getIDTokenWithCompletion:^(NSString * _Nullable token,
                                               NSError * _Nullable error) {
        if (error) {
            reject(@"TOKEN_ERROR", error.localizedDescription, error);
            return;
        }
        
        NSMutableDictionary *userData = [@{
            @"uid": authResult.user.uid,
            @"email": authResult.user.email ?: [NSNull null],
            @"displayName": authResult.user.displayName ?: [NSNull null],
            @"token": token,
            @"deviceFingerprint": self->deviceFingerprint,
            @"authTimestamp": @([[NSDate date] timeIntervalSince1970])
        } mutableCopy];
        
        // Add MFA status if available
        if ([authResult.user.multiFactor enrolledFactors].count > 0) {
            userData[@"mfaEnabled"] = @YES;
            userData[@"mfaEnrolledFactors"] = @([authResult.user.multiFactor enrolledFactors].count);
        }
        
        resolve(userData);
    }];
}

- (void)validateAndEnhanceUserData:(FIRUser *)user
                        withToken:(NSString *)token
                         resolve:(RCTPromiseResolveBlock)resolve
                          reject:(RCTPromiseRejectBlock)reject {
    
    // Verify token expiration
    FIRAuthTokenResult *decodedToken = [user getIDTokenResultForcingRefresh:NO
                                                                     error:nil];
    if (decodedToken.expirationDate.timeIntervalSinceNow < kTokenRefreshThreshold) {
        [user getIDTokenForcingRefresh:YES completion:^(NSString * _Nullable newToken,
                                                      NSError * _Nullable error) {
            if (error) {
                reject(@"TOKEN_REFRESH_ERROR", error.localizedDescription, error);
                return;
            }
            token = newToken;
        }];
    }
    
    NSMutableDictionary *enhancedUserData = [@{
        @"uid": user.uid,
        @"email": user.email ?: [NSNull null],
        @"displayName": user.displayName ?: [NSNull null],
        @"token": token,
        @"deviceFingerprint": deviceFingerprint,
        @"lastLoginTimestamp": @([[NSDate date] timeIntervalSince1970]),
        @"isEmailVerified": @(user.emailVerified)
    } mutableCopy];
    
    // Add claims and roles if available
    [user getIDTokenResultForcingRefresh:NO
                             completion:^(FIRAuthTokenResult * _Nullable tokenResult,
                                        NSError * _Nullable error) {
        if (!error && tokenResult.claims) {
            enhancedUserData[@"claims"] = tokenResult.claims;
            enhancedUserData[@"role"] = tokenResult.claims[@"role"] ?: @"user";
        }
        resolve(enhancedUserData);
    }];
}

- (void)logAuthError:(NSError *)error
           withEmail:(NSString *)email {
    NSDictionary *errorData = @{
        @"error_code": @(error.code),
        @"error_domain": error.domain,
        @"device_fingerprint": deviceFingerprint,
        @"masked_email": [self maskEmail:email]
    };
    // Log to analytics
    // Note: Actual logging implementation would use the Logger class
}

- (NSString *)maskEmail:(NSString *)email {
    if (!email.length) return @"";
    NSArray *components = [email componentsSeparatedByString:@"@"];
    if (components.count != 2) return @"";
    
    NSString *name = components[0];
    if (name.length <= 2) return email;
    
    return [NSString stringWithFormat:@"%@***%@@%@",
            [name substringToIndex:1],
            [name substringFromIndex:name.length - 1],
            components[1]];
}

+ (BOOL)requiresMainQueueSetup {
    return YES;
}

@end

File: src/ios/FantasyGMAssistant/NativeModules/Authentication/FirebaseAuthManager.swift
//
// FirebaseAuthManager.swift
// FantasyGMAssistant
//
// Enhanced Firebase Authentication manager with comprehensive security features
// Version: 1.0.0
//

import Foundation // iOS 14.0+
import FirebaseAuth // 10.0.0+
import GoogleSignIn // 7.0.0+
import React // 0.72+
import CryptoKit // iOS 14.0+

// MARK: - Constants
private let AUTH_ERROR_DOMAIN = "com.fantasygm.assistant.auth"
private let AUTH_QUEUE = DispatchQueue(label: "com.fantasygm.assistant.auth", qos: .userInitiated)
private let TOKEN_REFRESH_THRESHOLD: TimeInterval = 300 // 5 minutes before expiry
private let MAX_LOGIN_ATTEMPTS = 5
private let RATE_LIMIT_RESET: TimeInterval = 300 // 5 minutes

// MARK: - AuthError Enumeration
@objc enum AuthError: Int, Error {
    case invalidCredentials = 1001
    case networkError = 1002
    case userNotFound = 1003
    case emailInUse = 1004
    case weakPassword = 1005
    case mfaRequired = 1006
    case providerError = 1007
    case tokenExpired = 1008
    case rateLimitExceeded = 1009
    case suspiciousActivity = 1010
    case deviceNotTrusted = 1011
    
    var localizedDescription: String {
        switch self {
        case .invalidCredentials: return "Invalid email or password"
        case .networkError: return "Network connection error"
        case .userNotFound: return "User not found"
        case .emailInUse: return "Email already in use"
        case .weakPassword: return "Password does not meet security requirements"
        case .mfaRequired: return "Multi-factor authentication required"
        case .providerError: return "Authentication provider error"
        case .tokenExpired: return "Session expired, please login again"
        case .rateLimitExceeded: return "Too many login attempts, please try again later"
        case .suspiciousActivity: return "Suspicious activity detected"
        case .deviceNotTrusted: return "Device not trusted"
        }
    }
}

// MARK: - FirebaseAuthManager Class
@objc(FirebaseAuthManager)
class FirebaseAuthManager: NSObject {
    
    // MARK: - Properties
    private let auth: Auth
    private var currentUser: User?
    private let authQueue: DispatchQueue
    private var deviceFingerprint: String
    private var rateLimiter: [String: (attempts: Int, lastAttempt: Date)]
    private var tokenCache: [String: (token: String, expiry: Date)]
    
    // MARK: - Initialization
    override init() {
        self.auth = Auth.auth()
        self.authQueue = AUTH_QUEUE
        self.rateLimiter = [:]
        self.tokenCache = [:]
        
        // Generate device fingerprint
        let deviceData = "\(UIDevice.current.identifierForVendor?.uuidString ?? "")\(Bundle.main.bundleIdentifier ?? "")"
        self.deviceFingerprint = SHA256.hash(data: deviceData.data(using: .utf8)!).compactMap { String(format: "%02x", $0) }.joined()
        
        super.init()
        
        // Configure Firebase persistence
        do {
            try auth.useUserAccessGroup("group.com.fantasygm.assistant")
            auth.setPersistenceEnabled(true)
        } catch {
            Logger.shared.error("Failed to configure Firebase persistence", error: error)
        }
        
        // Setup auth state listener
        auth.addStateDidChangeListener { [weak self] (_, user) in
            guard let self = self else { return }
            self.currentUser = user
            if let user = user {
                Logger.shared.info("User authenticated: \(user.uid)")
                self.validateSession(user)
            } else {
                Logger.shared.info("User signed out")
                self.clearTokenCache()
            }
        }
    }
    
    // MARK: - Private Methods
    private func validateSession(_ user: User) {
        user.getIDTokenResult(forcingRefresh: false) { [weak self] (result, error) in
            guard let self = self else { return }
            
            if let error = error {
                Logger.shared.error("Token validation failed", error: error)
                self.signOut(nil, nil)
                return
            }
            
            guard let result = result else { return }
            
            // Validate claims and permissions
            guard let userRole = result.claims["role"] as? String,
                  ["free", "premium", "admin"].contains(userRole) else {
                Logger.shared.error("Invalid user role detected")
                self.signOut(nil, nil)
                return
            }
            
            // Check token expiration
            if result.expirationDate.timeIntervalSinceNow < TOKEN_REFRESH_THRESHOLD {
                self.refreshToken(nil, nil)
            }
        }
    }
    
    private func checkRateLimit(for identifier: String) -> Bool {
        return authQueue.sync {
            let now = Date()
            if let limitData = rateLimiter[identifier] {
                if limitData.attempts >= MAX_LOGIN_ATTEMPTS {
                    if now.timeIntervalSince(limitData.lastAttempt) < RATE_LIMIT_RESET {
                        return false
                    }
                    rateLimiter[identifier] = (0, now)
                }
                rateLimiter[identifier] = (limitData.attempts + 1, now)
            } else {
                rateLimiter[identifier] = (1, now)
            }
            return true
        }
    }
    
    private func clearTokenCache() {
        authQueue.async {
            self.tokenCache.removeAll()
        }
    }
    
    // MARK: - Public Methods
    @objc func signInWithEmail(_ email: String,
                              password: String,
                              resolve: @escaping RCTPromiseResolveBlock,
                              reject: @escaping RCTPromiseRejectBlock) {
        
        authQueue.async {
            // Rate limiting check
            guard self.checkRateLimit(for: email) else {
                Logger.shared.warning("Rate limit exceeded for email: [EMAIL]")
                reject("\(AuthError.rateLimitExceeded.rawValue)",
                      AuthError.rateLimitExceeded.localizedDescription,
                      AuthError.rateLimitExceeded)
                return
            }
            
            // Device trust check
            guard KeychainWrapper.standard.string(forKey: "device_trust_token") != nil else {
                Logger.shared.error("Untrusted device detected")
                reject("\(AuthError.deviceNotTrusted.rawValue)",
                      AuthError.deviceNotTrusted.localizedDescription,
                      AuthError.deviceNotTrusted)
                return
            }
            
            self.auth.signIn(withEmail: email, password: password) { (result, error) in
                if let error = error {
                    Logger.shared.error("Sign in failed", error: error)
                    let authError = self.handleAuthError(error)
                    reject("\(authError.rawValue)", authError.localizedDescription, authError)
                    return
                }
                
                guard let user = result?.user else {
                    reject("\(AuthError.userNotFound.rawValue)",
                          AuthError.userNotFound.localizedDescription,
                          AuthError.userNotFound)
                    return
                }
                
                // Get fresh token
                user.getIDTokenResult(forcingRefresh: true) { (result, error) in
                    if let error = error {
                        Logger.shared.error("Token retrieval failed", error: error)
                        reject("\(AuthError.tokenExpired.rawValue)",
                              AuthError.tokenExpired.localizedDescription,
                              AuthError.tokenExpired)
                        return
                    }
                    
                    guard let token = result?.token else { return }
                    
                    // Cache token
                    self.tokenCache[user.uid] = (token, result?.expirationDate ?? Date())
                    
                    // Return user data
                    let userData: [String: Any] = [
                        "uid": user.uid,
                        "email": user.email ?? "",
                        "emailVerified": user.isEmailVerified,
                        "token": token,
                        "expiresIn": result?.expirationDate.timeIntervalSinceNow ?? 0
                    ]
                    
                    Logger.shared.info("User signed in successfully: \(user.uid)")
                    resolve(userData)
                }
            }
        }
    }
    
    @objc func refreshToken(_ resolve: RCTPromiseResolveBlock?,
                           _ reject: RCTPromiseRejectBlock?) {
        authQueue.async {
            guard let user = self.auth.currentUser else {
                reject?("\(AuthError.userNotFound.rawValue)",
                       AuthError.userNotFound.localizedDescription,
                       AuthError.userNotFound)
                return
            }
            
            user.getIDTokenResult(forcingRefresh: true) { (result, error) in
                if let error = error {
                    Logger.shared.error("Token refresh failed", error: error)
                    reject?("\(AuthError.tokenExpired.rawValue)",
                           AuthError.tokenExpired.localizedDescription,
                           AuthError.tokenExpired)
                    return
                }
                
                guard let token = result?.token else { return }
                
                // Update token cache
                self.tokenCache[user.uid] = (token, result?.expirationDate ?? Date())
                
                let tokenData: [String: Any] = [
                    "token": token,
                    "expiresIn": result?.expirationDate.timeIntervalSinceNow ?? 0
                ]
                
                Logger.shared.info("Token refreshed successfully for user: \(user.uid)")
                resolve?(tokenData)
            }
        }
    }
    
    private func handleAuthError(_ error: Error) -> AuthError {
        let code = (error as NSError).code
        switch code {
        case AuthErrorCode.wrongPassword.rawValue:
            return .invalidCredentials
        case AuthErrorCode.userNotFound.rawValue:
            return .userNotFound
        case AuthErrorCode.emailAlreadyInUse.rawValue:
            return .emailInUse
        case AuthErrorCode.weakPassword.rawValue:
            return .weakPassword
        case AuthErrorCode.secondFactorRequired.rawValue:
            return .mfaRequired
        default:
            return .networkError
        }
    }
}